java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java

```diff
@@ -0,0 +1,848 @@
+   * and knowledge about what size to allocate is upto the callers of this
+   * method.
+   * Callers appropriately handle errors if memory allocation fails here.
+   * Callers should also take care of determining that desired size is
+   * within the bounds of max allocation allowed and any other error
+   * conditions.
+   */
+  private void allocateBytes(final long valueBufferSize, final long validityBufferSize) {
+      /* allocate data buffer */
+    int curSize = (int) valueBufferSize;
+    valueBuffer = allocator.buffer(curSize);
+    valueBuffer.readerIndex(0);
+    valueAllocationSizeInBytes = curSize;
+      /* allocate validity buffer */
+    allocateValidityBuffer((int) validityBufferSize);
+    zeroVector();
+  }
+
+  /**
+   * During splitAndTransfer, if we splitting from a random position within a byte,
+   * we can't just slice the source buffer so we have to explicitly allocate the
+   * validityBuffer of the target vector. This is unlike the databuffer which we can
+   * always slice for the target vector.
+   */
+  private void allocateValidityBuffer(final int validityBufferSize) {
+    validityBuffer = allocator.buffer(validityBufferSize);
+    validityBuffer.readerIndex(0);
+    validityAllocationSizeInBytes = validityBufferSize;
+  }
+
+  /**
+   * Get the potential buffer size for a particular number of records.
+   * @param count desired number of elements in the vector
+   * @return estimated size of underlying buffers if the vector holds
+   *         a given number of elements
+   */
+  @Override
+  public int getBufferSizeFor(final int count) {
+    if (count == 0) {
+      return 0;
+    }
+    return (count * typeWidth) + getValidityBufferSizeFromCount(count);
+  }
+
+  /**
+   * Get the size (number of bytes) of underlying buffers used by this
+   * vector
+   * @return size of underlying buffers.
+   */
+  @Override
+  public int getBufferSize() {
+    if (valueCount == 0) {
+      return 0;
+    }
+    return (valueCount * typeWidth) + getValidityBufferSizeFromCount(valueCount);
+  }
+
+  /**
+   * Get information about how this field is materialized.
+   * @return the field corresponding to this vector
+   */
+  @Override
+  public Field getField() {
+    return field;
+  }
+
+  /**
+   * Return the underlying buffers associated with this vector. Note that this doesn't
+   * impact the reference counts for this buffer so it only should be used for in-context
+   * access. Also note that this buffer changes regularly thus
+   * external classes shouldn't hold a reference to it (unless they change it).
+   *
+   * @param clear Whether to clear vector before returning; the buffers will still be refcounted
+   *              but the returned array will be the only reference to them
+   * @return The underlying {@link io.netty.buffer.ArrowBuf buffers} that is used by this
+   *         vector instance.
+   */
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
```

Comment:
My previous comments are squashed because of code change.

Why do we need both 

```
public ArrowBuf[] getBuffers(boolean clear)
``` 
and 
```
public List<ArrowBuf> getFieldBuffers()
```
Both as public API? They look very similar to me and might confuse users.

Reply:
AFAIK, getBuffers() is an old method for getting a batch of buffers for serialization. We use it for maintaining compatibility with Drill client. I think in future, we should consider removing it from Arrow and writing a wrapper over it in Dremio. This way we will just have getFieldBuffers.

Reply:
I see. Does it make sense to mark getBuffers() as deprecated so it's less confusing for the other users?

Reply:
Let's not worry about it right now. These two APIs have not been introduced as part of refactoring and have been in Arrow code for quite some time. 

I will file a JIRA and we can brainstorm w.r.t appropriate steps to address the ambiguity (if any)

Reply:
Ok. Sounds good.